plugins/timertop/timermodel.cpp

```diff
@@ -27,190 +27,387 @@
+        changed = true;
+    }
+
+    TimerIdInfo &toInfo(TimerId::Type type)
+    {
+        TimerIdInfo::totalWakeups =  TimerIdData::totalWakeups(type);
+        TimerIdInfo::wakeupsPerSec = TimerIdData::wakeupsPerSec(type);
+        TimerIdInfo::timePerWakeup = TimerIdData::timePerWakeup(type);
+        TimerIdInfo::maxWakeupTime = TimerIdData::maxWakeupTime(type);
+        return *this;
+    }
+
+    int totalWakeupsEvents;
+    FunctionCallTimer functionCallTimer;
+    QList<TimeoutEvent> timeoutEvents;
+    quintptr lastReceiver; // free timers only
+
+    bool changed;
+};
+}
+
+Q_DECLARE_METATYPE(GammaRay::TimeoutEvent)
 
 TimerModel::TimerModel(QObject *parent)
     : QAbstractTableModel(parent)
     , m_sourceModel(nullptr)
-    , m_pendingChanedRowsTimer(new QTimer(this))
     , m_timeoutIndex(QTimer::staticMetaObject.indexOfSignal("timeout()"))
     , m_qmlTimerTriggeredIndex(-1)
+    , m_pushTimerStart(QTimer::staticMetaObject.method(QTimer::staticMetaObject.indexOfSlot("start()")))
 {
-    m_pendingChanedRowsTimer->setInterval(5000);
-    m_pendingChanedRowsTimer->setSingleShot(true);
-    connect(m_pendingChanedRowsTimer, SIGNAL(timeout()), this, SLOT(flushEmitPendingChangedRows()));
-}
+    qRegisterMetaType<TimerId>();
+    qRegisterMetaType<TimerIdInfo>();
+    qRegisterMetaType<TimerIdInfoHash>();
+    qRegisterMetaType<GammaRay::TimeoutEvent>();
 
-TimerModel::~TimerModel()
-{
-    s_timerModel = nullptr;
+    const int applyChangesIndex = TimerModel::staticMetaObject.indexOfSlot("applyChanges(GammaRay::TimerIdInfoHash)");
+    m_applyChangesMethod = TimerModel::staticMetaObject.method(applyChangesIndex);
+    Q_ASSERT(m_applyChangesMethod.methodIndex() != -1);
+
+    QInternal::registerCallback(QInternal::EventNotifyCallback, eventNotifyCallback);
 }
 
-bool TimerModel::isInitialized()
+const TimerIdInfo *TimerModel::findTimerInfo(const QModelIndex &index) const
 {
-    return s_timerModel != nullptr;
+    if (index.row() < m_sourceModel->rowCount()) {
+        const QModelIndex sourceIndex = m_sourceModel->index(index.row(), 0);
+        QObject * const timerObject = sourceIndex.data(ObjectModel::ObjectRole).value<QObject *>();
+        const TimerId id = TimerId(timerObject);
+        auto it = m_timersInfo.find(id);
+
+        if (it == m_timersInfo.end()) {
+            it = m_timersInfo.insert(id, TimerIdInfo(id));
+        }
+
+        if (it != m_timersInfo.end())
+            return &it.value();
+    } else {
+        if (index.row() < m_sourceModel->rowCount() + m_freeTimersInfo.count())
+            return &m_freeTimersInfo[index.row() - m_sourceModel->rowCount()];
+    }
+
+    return nullptr;
 }
 
-TimerModel *TimerModel::instance()
+void TimerModel::triggerPushChanges()
 {
-    if (!s_timerModel)
-        s_timerModel = new TimerModel;
-    Q_ASSERT(s_timerModel);
-    return s_timerModel;
+    if (!TimerModel::isInitialized())
+        return;
+
+    QMutexLocker locker(Probe::objectLock());
```

Comment:
Why are we holding the object lock here, we are not checking QObject validity anywhere from what I can see?